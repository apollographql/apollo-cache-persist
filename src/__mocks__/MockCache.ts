import Cache from '../Cache';
import { ApolloPersistOptions, PersistedData } from '../types';

export default class MockCache<T, U extends boolean = true>
  implements Cache<T>
{
  cache: null;
  serialize: boolean;
  data: PersistedData<T>;

  constructor(options: Pick<ApolloPersistOptions<T, U>, 'serialize'>) {
    const { serialize = true } = options;
    this.serialize = serialize;
  }

  extract(): PersistedData<T> {
    let data: PersistedData<T> = this.data;

    if (this.serialize) {
      data = JSON.stringify(this.data) as string;
    }

    return data;
  }

  restore(data: PersistedData<T>): void {
    if (this.serialize && typeof data === 'string') {
      data = JSON.parse(data);
    }

    if (data != null) {
      this.data = data;
    }
  }
}
