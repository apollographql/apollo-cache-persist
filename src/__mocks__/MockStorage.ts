import { PersistentStorage } from '../types';

export default class MockStorage implements PersistentStorage {
  storage: Map<string, string>;

  constructor() {
    this.storage = new Map();
  }

  setItem(key: string, data: string): Promise<any> {
    return new Promise(resolve => {
      this.storage.set(key, data);
      resolve();
    });
  }

  removeItem(key: string): Promise<any> {
    return new Promise((resolve, reject) => {
      this.storage.delete(key);
      resolve();
    });
  }

  getItem(key: string): Promise<string> {
    return new Promise((resolve, reject) => {
      resolve(this.storage.get(key));
    });
  }
}
