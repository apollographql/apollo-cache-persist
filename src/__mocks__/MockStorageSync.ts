import { PersistentStorage } from '../types';

export default class MockStorageSync implements PersistentStorage {
  storage: Map<string, string>;

  constructor() {
    this.storage = new Map();
  }

  setItem(key: string, data: string): void {
    this.storage.set(key, data);
  }

  removeItem(key: string): void {
    this.storage.delete(key);
  }

  getItem(key: string): any {
    return this.storage.get(key);
  }
}
