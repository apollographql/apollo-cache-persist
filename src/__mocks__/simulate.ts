import { ApolloClient } from 'apollo-client';
import { ApolloLink, Observable } from 'apollo-link';
import { InMemoryCache } from 'apollo-cache-inmemory';
import gql from 'graphql-tag';

import { persistCache } from '../';
import MockStorage from './MockStorage';

export const simulateApp = async ({
  result,
  operation,
  persistOptions = {},
}) => {
  const cache = persistOptions.cache || new InMemoryCache();
  const storage = new MockStorage();

  await persistCache({ storage, ...persistOptions, cache });

  const link = new ApolloLink(() => {
    return Observable.of(result);
  });
  const client = client || new ApolloClient({ cache, link });

  await client.query({ query: operation });
  jest.runTimersToTime(
    persistOptions.debounce ? persistOptions.debounce + 1 : 1001
  );
  const extracted = client.extract();

  // cache is now persisted
  const cache2 = new cache.constructor();
  await persistCache({ cache: cache2, storage });
  const client2 = new ApolloClient({ cache: cache2, link });

  return [client, client2];
};

export const simulateWrite = async ({
  cache = new InMemoryCache(),
  storage = new MockStorage(),
  result,
  operation,
  ...rest,
}) => {
  await persistCache({ storage, cache, ...rest });

  const link = new ApolloLink(() => {
    return Observable.of(result);
  });
  const client = new ApolloClient({ cache, link });
  await client.query({ query: operation });
};
