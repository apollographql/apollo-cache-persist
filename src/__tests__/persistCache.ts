import gql from 'graphql-tag';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { Hermes } from 'apollo-cache-hermes';

import { persistCache } from '../';
import MockStorage from '../__mocks__/MockStorage';
import { simulateApp, simulateWrite } from '../__mocks__/simulate';

jest.useFakeTimers();
describe('persistCache', () => {
  describe('setup', () => {
    it('requires a cache', async () => {
      try {
        await persistCache({ storage: new MockStorage() });
        fail('invoking persistCache without a cache should throw an error');
      } catch (e) {
        expect(() => {
          throw e;
        }).toThrowErrorMatchingSnapshot();
      }
    });
    it('requires storage', async () => {
      try {
        await persistCache({ cache: new InMemoryCache() });
        fail('invoking persistCache without storage should throw an error');
      } catch (e) {
        expect(() => {
          throw e;
        }).toThrowErrorMatchingSnapshot();
      }
    });
  });

  describe('basic usage', () => {
    const operation = gql`
      {
        hello
      }
    `;
    const result = { data: { hello: 'world' } };

    it('extracts a previously filled InMemoryCache from storage', async () => {
      const [client, client2] = await simulateApp({
        operation,
        result,
      });
      expect(client.extract()).toEqual(client2.extract());
    });
    xit('extracts a previously filled HermesCache from storage', async () => {
      const [client, client2] = await simulateApp({
        operation,
        result,
        persistOptions: {
          cache: new Hermes(),
        },
      });
      expect(client.extract()).toEqual(client2.extract());
    });
  });

  describe('nested queries', () => {
    const operation = gql`
      {
        user(id: 1) {
          name {
            last
            first
          }
          posts {
            title
            comments {
              name
            }
          }
        }
      }
    `;
    const result = {
      data: {
        user: {
          name: { last: 'Doe', first: 'Jane', __typename: 'Name' },
          posts: [
            {
              title: 'Apollo is awesome',
              comments: [{ name: 'foo', __typename: 'Comment' }],
              __typename: 'Post',
            },
          ],
          __typename: 'User',
        },
      },
    };

    it('extracts a previously filled InMemoryCache from storage', async () => {
      const [client, client2] = await simulateApp({
        operation,
        result,
      });
      expect(client.extract()).toEqual(client2.extract());
    });
    xit('extracts a previously filled HermesCache from storage', async () => {
      const [client, client2] = await simulateApp({
        operation,
        result,
        persistOptions: {
          cache: new Hermes(),
        },
      });
      expect(client.extract()).toEqual(client2.extract());
    });
  });

  describe('advanced usage', () => {
    const operation = gql`
      {
        hello
      }
    `;
    const result = { data: { hello: 'world' } };

    it('passing in debounce configures the debounce interval', async () => {
      const debounce = 600;
      const storage = new MockStorage();

      await simulateWrite({
        result,
        operation,
        persistOptions: { debounce, storage },
      });

      expect(await storage.getItem('apollo-cache-persist')).toBe(undefined);
      jest.runTimersToTime(debounce + 1);
      expect(await storage.getItem('apollo-cache-persist')).toMatchSnapshot();
    });
    it('passing in key customizes the storage key', async () => {
      const storage = new MockStorage();
      const key = 'testing-1-2-3';

      await simulateWrite({
        result,
        operation,
        persistOptions: { key, storage },
      });

      jest.runTimersToTime(1000);
      expect(await storage.getItem('apollo-cache-persist')).toBe(undefined);
      expect(await storage.getItem(key)).toMatchSnapshot();
    });
    xit('setting the trigger to background does not persist on a write', async () => {
      const storage = new MockStorage();
      await simulateWrite({
        result,
        operation,
        persistOptions: { trigger: 'background', storage },
      });

      jest.runTimersToTime(1000);
      expect(await storage.getItem('apollo-cache-persist')).toBe(undefined);
    });
    xit('setting the trigger to background persists in the background');
  });
});
