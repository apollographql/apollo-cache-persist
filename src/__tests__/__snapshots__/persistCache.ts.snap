// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`persistCache advanced usage passing in debounce configures the debounce interval 1`] = `"{\\"ROOT_QUERY\\":{\\"hello\\":\\"world\\"}}"`;

exports[`persistCache advanced usage passing in key customizes the storage key 1`] = `"{\\"ROOT_QUERY\\":{\\"hello\\":\\"world\\"}}"`;

exports[`persistCache setup requires a cache 1`] = `
"
      In order to persist your Apollo Cache, you need to pass in a cache.
      Please see https://www.apollographql.com/docs/react/basics/caching.html for our default InMemoryCache.
    "
`;

exports[`persistCache setup requires storage 1`] = `
"
      In order to persist your Apollo Cache, you need to pass in an underlying storage provider.
      Please see https://github.com/apollographql/apollo-cache-persist#storage-providers
    "
`;
